samcli/lib/sync/infra_sync_executor.py

```diff
@@ -0,0 +1,284 @@
+        self._cfn_client = self._boto_client("cloudformation")
+        self._s3_client = self._boto_client("s3")
+
+    def _boto_client(self, client_name: str):
+        """
+        Creates boto client
+
+        Parameters
+        ----------
+        client_name: str
+            The name of the client
+
+        Returns
+        -------
+        Service client instance
+        """
+        return get_boto_client_provider_from_session_with_config(self._session)(client_name)
+
+    def _compare_templates(self, local_template_path: str, stack_name: str) -> bool:
+        """
+        Recursively conpares two templates, including the nested templates referenced inside
+
+        Parameters
+        ----------
+        local_template_path : str
+            The local template location
+        stack_name : str
+            The CloudFormation stack name that the template is deployed to
+
+        Returns
+        -------
+        bool
+            Returns True if two templates are identical
+            Returns False if two templates are different
+        """
+        current_template = None
+        # If the customer template uses a nested stack with location/template URL in S3
+        if local_template_path.startswith("https://"):
+            parsed_s3_location = re.search(r"https:\/\/[^/]*\/([^/]*)\/(.*)", local_template_path)
+            if parsed_s3_location:
+                s3_bucket = parsed_s3_location.group(1)
+                s3_key = parsed_s3_location.group(2)
+                try:
+                    s3_object = self._s3_client.get_object(Bucket=s3_bucket, Key=s3_key)
+                except ClientError as ex:
+                    LOG.debug("The provided template location %s can not be found", local_template_path, exc_info=ex)
+                    return False
+
+                streaming_body = s3_object.get("Body")
+                if streaming_body:
+                    current_template = yaml_parse(streaming_body.read().decode("utf-8"))
+
+        # If the template location is local
+        else:
+            current_template = get_template_data(local_template_path)
+
+        if not current_template:
+            return False
+
+        try:
+            last_deployed_template_str = self._cfn_client.get_template(
+                StackName=stack_name, TemplateStage="Original"
+            ).get("TemplateBody", "")
+        except ClientError as ex:
+            LOG.debug("Stack with name %s does not exist on CloudFormation", stack_name, exc_info=ex)
+            return False
+
+        last_deployed_template_dict = yaml_parse(last_deployed_template_str)
+
+        sanitized_resources = self._remove_unnecessary_fields(current_template)
+        self._remove_unnecessary_fields(last_deployed_template_dict, sanitized_resources)
+
+        if last_deployed_template_dict != current_template:
+            return False
```

Comment:
Please add some debugging log here.

Reply:
Next PR: if we decide that the comparison happens before packaging, then this logic needs to be improved because nested stack resources are not sanitized, and it's possible for a early False return due to the template comparison here before going down to the recursive check below, which defeats partially the purpose below.

Reply:
The comparison will happen after packaging, but have addressed the above concern by making a deepcopy for sanitization, also having nested stack resources as a part of sanitized resources